src/Compilers/CSharp/Test/Semantic/Semantics/GenericConstraintsTests.cs

```diff
@@ -3629,5 +3629,307 @@ public struct YourStruct<T> where T : unmanaged
+        }
+
+        [Fact]
+        public void SimpleGenericStructPointer_ILValidation()
+        {
+            var code = @"
+public unsafe struct MyStruct<T> where T : unmanaged
+{
+    public T field;
+
+    public static void Test()
+    {
+        var ms = new MyStruct<int>();
+        MyStruct<int>* ptr = &ms;
+        ptr->field = 42;
+    }
+}
+";
+            var il = @"
+{
+  // Code size       19 (0x13)
+  .maxstack  2
+  .locals init (MyStruct<int> V_0) //ms
+  IL_0000:  ldloca.s   V_0
+  IL_0002:  initobj    ""MyStruct<int>""
+  IL_0008:  ldloca.s   V_0
+  IL_000a:  conv.u
+  IL_000b:  ldc.i4.s   42
+  IL_000d:  stfld      ""int MyStruct<int>.field""
+  IL_0012:  ret
+}
+";
+            CompileAndVerify(code, options: TestOptions.UnsafeReleaseDll, verify: Verification.Skipped)
+                .VerifyIL("MyStruct<T>.Test", il);
+        }
+
+        [Fact]
+        public void CircularGenericUnmanagedInstantation()
+        {
+            var code = @"
+public struct X<T>
+    where T : unmanaged
+{
+}
+
+public struct Z
+{
+    public X<Z> field;
+}";
+            CreateCompilation(code).VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void GenericStructAddressOfRequiresCSharp8()
+        {
+            var code = @"
+public struct MyStruct<T>
+{
+    public T field;
+
+    public static unsafe void Test()
+    {
+        var ms = new MyStruct<int>();
+        var ptr = &ms;
+    }
+}
+";
+            CreateCompilation(code, options: TestOptions.UnsafeReleaseDll, parseOptions: TestOptions.Regular7_3)
+                .VerifyDiagnostics(
+                    // (9,19): error CS8370: Feature 'unmanaged generic structs' is not available in C# 7.3. Please use language version 8.0 or greater.
+                    //         var ptr = &ms;
+                    Diagnostic(ErrorCode.ERR_FeatureNotAvailableInVersion7_3, "&ms").WithArguments("unmanaged generic structs", "8.0").WithLocation(9, 19)
+                );
+        }
+
+        [Fact]
+        public void GenericStructFixedRequiresCSharp8()
+        {
+            var code = @"
+public struct MyStruct<T>
+{
+    public T field;
+}
+
+public class MyClass
+{
+    public MyStruct<int> ms;
+    public static unsafe void Test(MyClass c)
+    {
+        fixed (MyStruct<int>* ptr = &c.ms)
+        {
+        }
+    }
+}
+";
+            CreateCompilation(code, options: TestOptions.UnsafeReleaseDll, parseOptions: TestOptions.Regular7_3)
+                .VerifyDiagnostics(
+                    // (12,16): error CS8370: Feature 'unmanaged generic structs' is not available in C# 7.3. Please use language version 8.0 or greater.
+                    //         fixed (MyStruct<int>* ptr = &c.ms)
+                    Diagnostic(ErrorCode.ERR_FeatureNotAvailableInVersion7_3, "MyStruct<int>*").WithArguments("unmanaged generic structs", "8.0").WithLocation(12, 16)
```

Comment:
This test produces a redundant diagnostic. e.g.

```cs
// (12,16): error CS8370: Feature 'unmanaged generic structs' is not available in C# 7.3. Please use language version 8.0 or greater.
//         fixed (MyStruct<int>* ptr = &c.ms)
Diagnostic(ErrorCode.ERR_FeatureNotAvailableInVersion7_3, "MyStruct<int>*").WithArguments("unmanaged generic structs", "8.0").WithLocation(12, 16),
// (12,37): error CS8370: Feature 'unmanaged generic structs' is not available in C# 7.3. Please use language version 8.0 or greater.
//         fixed (MyStruct<int>* ptr = &c.ms)
Diagnostic(ErrorCode.ERR_FeatureNotAvailableInVersion7_3, "&c.ms").WithArguments("unmanaged generic structs", "8.0").WithLocation(12, 37)
```

Not sure what I should do about it yet.

Reply:
I think that's expected here. There are two uses of the new feature: the type declaration and the address-of argument. Both should be issuing a diagnostic. 

---
In reply to: [243082608](https://github.com/dotnet/roslyn/pull/31602#discussion_r243082608) [](ancestors = 243082608)

Reply:
OK, I'll probably move forward by just pasting in those expected diagnostics.

Reply:
Agreed.

Reply:
Done